MA-1190; Thread PUT -  update thread read status

wajeeha-khalid opened the pull request:
MA-1190; Thread PUT -  update thread read status

wajeeha-khalid commented:
@nasthagiri @BenjiLee kindly review

nasthagiri commented:
@wajeeha-khalid Is there a corresponding change on the server?  For example, (1) updating the docstring for PATCH and (2) removing the `mark_as_read` parameter in the GET method?

nasthagiri commented:
@cahrens Is there someone on your team who can review the Ruby code?  Or should we pull in Bill?

wajeeha-khalid commented:
@nasthagiri Yes, There is Patch change at mobile API side; but i am waiting on other two PRs for thread patch to get merged so I can create PR for this change. I have made the changes already.

cahrens commented on spec/api/comment_thread_spec.rb:
```diff
@@ -594,6 +594,24 @@ def thread_result(id, params)
+        changed_thread.commentable_id.should == "new_commentable_id"
+        changed_thread.thread_type.should == "question"
+        comment.reload
+        comment.endorsed.should == false
```
Why is endorsed now set to false?

And where in this test does it check that the thread is now marked to "read"?

cahrens commented:
I'm unclear if the test is covering the API code change. But the change to the API itself looks fine.

@robrap is also going to review.

robrap commented on api/comment_threads.rb:
```diff
@@ -54,10 +54,15 @@
 
+  if params["user_id"] and value_to_boolean(params["read"])
+    user = User.only([:id, :username, :read_states]).find_by(external_id: params["user_id"])
+    user.mark_as_read(thread) if user
```
if we have no user at this point, wouldn't we want some sort of exception or error logged to indicate something unexpected is happening?

wajeeha-khalid replied:
@robrap Comment service is being used by multiple clients and initially there was no check for `user_id` or `read` params. This way it will simply ignore this part if these params are not passed to the endpoint; if we start throwing exception it might crash other clients. So i have tried to keep the default behaviour as is. Also as we are planning to rebuild comment service I believe we should not bring new changes for existing client for now.

robrap replied:
@wajeeha-khalid.  Agreed we should not add in the exception.  And it seems the comment service doesn't have a logging mechanism for logging the unexpected (e.g. user id is passed but user can't be found).  I'll document this for future comment service improvements.  Thanks.

robrap commented on spec/api/comment_thread_spec.rb:
```diff
@@ -594,6 +594,24 @@ def thread_result(id, params)
+        comment.reload
+        comment.endorsed.should == false
+        comment.endorsement.should == nil
+        check_thread_result_json(nil, changed_thread, parse(last_response.body))
```
I agree with @cahrens.  This test seems the same as the one above it where we don't send read. 
* Does the first param of check_thread_result_json need to be the user?  Without this, it seems it is just checking for read to be false, but maybe I'm misreading the code.
* This test and the test above seem exactly the same except for the read param.
  * check_thread_result_json check seems to check that things are consistent, but not that they are accurate.  There should be a separate explicit assertion here for read to be true.
  * I'd change the earlier test to be named "update information of comment thread and don't mark thread as read" for consistency.  It should have an explicit assertion for read being false.

wajeeha-khalid commented:
@cahrens @robrap `check_thread_result_json` confirms the behavior for `read'; my bad that i didnot pass `user' to it which was equivalant to skip that check. Updated PR with correction also added explicit assertion as @robrap suggested.

wajeeha-khalid commented:
@cahrens @robrap kindly give a second round review
We have https://github.com/edx/edx-platform/pull/10649 dependent on this

cahrens commented on spec/api/comment_thread_spec.rb:
```diff
@@ -592,7 +592,24 @@ def thread_result(id, params)
+        changed_thread.title.should == "new title"
+        changed_thread.commentable_id.should == "new_commentable_id"
+        changed_thread.thread_type.should == "question"
+        user.mark_as_read(thread)
```
I am still confused about this test. "read" is marked on line 602. Why user.mark_as_read(thread) on line 609?

robrap replied:
Agreed.  If you need line 609 to make this pass it means there is likely a bug, even if it is only in the check_thread_result_json assertion code.

wajeeha-khalid replied:
@cahrens @robrap 
That was because the user is created within this test and does not get updated with read states (as it is not returned in response of API where read_states are being added). So I replicated the behaviour of mark_as_read from API to test (as it was being used in all GET tests so I assumed we are using it as convention). But you guys are right that this makes confusion so instead of replicating mark_ad_read I have reloaded user to get its read_states.

Updated this PR; kindly review
P.S:
Should we consider changing GET test cases too? If yes, then probably we can create a task in backlog for that.

robrap commented on spec/api/comment_thread_spec.rb:
```diff
@@ -592,7 +592,24 @@ def thread_result(id, params)
         comment.endorsement.should == nil
-        check_thread_result_json(nil, changed_thread, parse(last_response.body))
+        json_response = parse(last_response.body)
+        check_thread_result_json(nil, changed_thread, json_response)
```
@wajeeha-khalid In your comment you mention the GET cases.  I imagine they should match.  It seems like it should be more clear where you would ever want check_thread_result_json with user of nil, as opposed to getting the user.

At a minimum, these 2 update tests should match.  In other words, we should get the user and sent it in here.

wajeeha-khalid replied:
@robrap 
In that case I am wondering why `user.nil` case has been put there on first place (https://github.com/edx/cs_comments_service/blob/master/spec/spec_helper.rb#L258)
May be to improve performance; as according to API if user is not provided thread cannot be marked as read. Same goes with the test. As if thread is not marked as read then instead of making an extra query to get user `read_states` and checking its properties, it explicitly asserts `read` as `false` with user=nil.
What do you think?

cahrens commented on spec/spec_helper.rb:
```diff
@@ -282,6 +282,10 @@ def check_thread_result_json(user, thread, json_response)
   check_thread_result(user, thread, json_response, true)
 end
 
+def check_unread_thread_result_json(thread, json_response)
```
Please add a comment to explain why nil is passed as the user.

cahrens commented:
:+1: after you add the comment and get the tests passing

robrap commented:
:+1:
